### DISCUSSION TITLE
A workaround for retrying assertions on network requests

### DISCUSSION THREAD

**Discussion by msurdi** [cypress-io/cypress#19118] [Questions and Help]
Hi!

I've recently had to solve a problem where I needed to test the requests sent to an external service by a third party library in my application. The purpose of the library is to track certain events, so I wanted to assert the list of events such as "event 1", "event 2", etc that were triggered after a given interaction in the application.

The main difficulty I found was that those requests that are sending the events to the third party service, due to the batching and throttling the library does internally, were triggered in a totally unpredictable number, at unpredictable times and with unpredictable payload each one of them.

So far, the solution I've used in these cases was to just put a long enough sleep, like `cy.wait(10000)` and then grab all the events from all the network requests (with `cy.get("@trackingCalls.all")` join them on a list and then assert this list was the expected ones. But the problem of this approach is that we have to wait for 10s on each test, which is very undersirable.

So after giving another try to the problem, I came up with the following solution. I wanted to share here and also ask:

1. Do you think there is a better / more idiomatic way to accomplish this?
2. Do you think this can have any undesired side effect I'm not aware of?

**The solution**:

```JavaScript
let events = []
const getEvents = () => events
const resetEvents = () => { events = [] }
const registerEvents = (newEvents) => { events.push(...newEvents) }

export const mockTracking = () => {
  cy.intercept({
    method: 'POST',
    path: '/track/*'
  }, (req) => {
    resetEvents()
      const newEvents = req?.body?.events
      registerEvents(newEvents)
    req.reply(
      {
        statusCode: 200,
        body: { ok: true }
      }
    )
  })
}

const markAsDone = () => {
  const div = document.createElement('div')
  div.setAttribute('id', 'tracking-ready')
  div.style = { display: 'none' }
  document.body.appendChild(div)
}

export const testTrackingEvents = (expectedEvents) => {
  cy.get('#tracking-ready', { timeout: 10000 }).should(() => {
    const currentEvents = getEvents()
    if (currentEvents.length === expectedEvents.length) {
      markAsDone()
    }
    expect(currentEvents.sort()).to.deep.eq(expectedEvents.sort())
  })
}
```

**Usage**:

```JavaScript
it("should send events", () => {
  mockTracking()
  // Do some interactions
  testTrackingEvents(["event 1", "event 2"])
})
```

So basically given cypress automatically retries assertions on DOM elements, I make it wait for a DOM element which initially does not exist, but then on each try I can test if the number of events I'm expecting have been already submitted, if so then we create the DOM element. In the next retry the DOM element will be present so the real assertion can be performed.

This wait the 10s wait becomes the worst case scenario (when the events don't match) but when all is good the test will take the least necessary amount of time.

I know, this is a really dirty hack 😃 . What do you think?

Thanks!
